content/posts/redis-modules-into-arm-land-part1.md

```diff
@@ -0,0 +1,275 @@
+Finally, [**RedisAI**](https://github.com/RedisAI/RedisAI). It's a C library built with CMake, and it includes modular "engines" that allow abstraction and encapsulation of AI libraries like [TensorFlow](https://github.com/tensorflow/tensorflow), [PyTorch](https://github.com/pytorch/pytorch), and [ONNXRuntime](https://github.com/microsoft/onnxruntime), in their C library form (most users typically use them in Python), with PyTorch and ONNXRuntime not officially supporting ARM.
+So the build requirements, as it seems, have deteriorated quickly. It went from building an innocent C library to compiling massive source bases with convoluted build systems.
+
+In the next sections, we will fit each components with its proper build method.
+
+## Building for ARM
+
+At this point, we'll pause and take stock of what is required to build software for ARM.
+The obvious way is to use an ARM-based device, like Raspberry Pi. Once set up, you'll be able to build and test in the most natural manner. 
+
+The testing medium is important: even though you can build almost every ARM software without a physical ARM device, there is no reliable way of testing the outcome in a reliable way without such a device, especially with non-standard devices. Therefore, as much as virtualized/emulated/containerized test may be useful, you should always test your software on its designated target device.
+
+A Raspberry Pi 4 with 4GB RAM (and not less importantly, a 1Gbps NIC) and a fast microSD card looks indeed very promising.
+
+Now, regarding the OS selection on ARM: the offerings are limited, and the rule of thumb is going for the latest release of either Raspbian (which is customized Debian distribution for RPi) or either Ubuntu or Fedora, both offering easy-to-install ARM systems. Do not worry about immaturity: it was proven time
+and again that newer systems work better and old ones can't keep up.
+
+Installing an OS brings us to the first dilemma: 
+If we install Raspbian, we get a 32-bit OS (i.e., arm32v7 platform).
+If we choose 64-bit Ubuntu, we get an arm64v8 platform (we discuss ARM platforms in detail later). 
+If we intend (as we do with RedisEdge) to support both platforms, we can either get two SD cards, and install each OS on its own card, taking turns on the device, or get two RPi devices (which is not terribly expensive). I recommend the latter.
+
+And now, for the principal principle of OS selection: our goal is to have a stable system with the newest Docker version. Any OS that will satisfy these requirements will do.
+That's right: we're going to use Docker to fetch the OS we really with to build for, using the underlying OS as infrastructure. This will also help keep our build experiments properly isolated.

```

Comment:
Is there a word missing between "really and with"? Additionally, change to: "That's right—we're going to use Docker to fetch the OS we really with to build for, using the underlying OS as infrastructure. This will also help keep our build experiments properly isolated."

Reply:
"with" is a typo, I meant "wish".